sympy/physics/control/lti.py

```diff
@@ -525,16 +547,88 @@ def doit(self, **kwargs):
             if res is None:
                 res = arg
             else:
-                num_ = arg.num * res.num
-                den_ = arg.den * res.den
-                res = TransferFunction(num_, den_, self.var)
+                if self.is_SISO:
+                    num_ = arg.num * res.num
+                    den_ = arg.den * res.den
+                    res = TransferFunction(num_, den_, self.var)
+                else:
+                    # Now we multiply two transfer function matrices...
+                    # If we have two TFMs with shape (a, b) and (b, d), respectively,
+                    # then the resultant TFM will be of shape (a, d).
+                    if res.num_outputs == 1 and arg.num_inputs == 1:
+                        a = [None]
+                        for i in range(res.num_inputs):
+                            if a[0] is None:
+                                a[0] = res.args[0][0][i] * arg.args[0][i]
+                            else:
+                                a[0] += res.args[0][0][i] * arg.args[0][i]
+
+                    elif res.num_outputs == 1 or arg.num_inputs == 1:
+                        if res.num_outputs == 1:
+                            a = [[None] * arg.num_inputs]
+
+                            for j in range(arg.num_inputs):
+                                for k in range(arg.num_outputs):
+                                    if a[0][j] is None:
+                                        a[0][j] = res.args[0][0][j] * arg.args[0][k][j]
+                                    else:
+                                        a[0][j] += res.args[0][0][j] * arg.args[0][k][j]
+
+                        else:
+                            a = [None] * res.num_outputs
+
+                            for i in range(res.num_outputs):
+                                for k in range(arg.num_outputs):
+                                    if a[i] is None:
+                                        if res.num_inputs == 1:
+                                            a[i] = res.args[0][i] * arg.args[0][k]
+                                        else:
+                                            a[i] = res.args[0][i][k] * arg.args[0][k]
+                                    else:
+                                        a[i] += res.args[0][i][k] * arg.args[0][k]
+
+                    else:
```

Comment:
Is there any reason to have these special cases? I don't think trying to optimize out `for i in range(1)` is worthwhile.

Reply:
Thanks! I have those special cases for several reasons.
 For instance: consider `tfm1` with shape (2, 3), `tfm2` with shape (3, 1), and `tfm3` with shape (2, 1). I get the following traceback:
``` python
________________________________________________________________________ sympy/physics/control/tests/test_lti.py:test_Series_functions ________________________________________________________________________
Traceback (most recent call last):
  File "/home/namannimmo/sympy/sympy/physics/control/tests/test_lti.py", line 512, in test_Series_functions
    assert tfm1*tfm2 + tfm6 == Parallel(Series(tfm1, tfm2), tfm6)
  File "/home/namannimmo/sympy/sympy/physics/control/lti.py", line 607, in __add__
    return Parallel(self, other)
  File "/home/namannimmo/sympy/sympy/physics/control/lti.py", line 826, in __new__
    obj.is_SISO = True if all(isinstance(arg.doit(), TransferFunction) for arg in args) \
  File "/home/namannimmo/sympy/sympy/physics/control/lti.py", line 826, in <genexpr>
    obj.is_SISO = True if all(isinstance(arg.doit(), TransferFunction) for arg in args) \
  File "/home/namannimmo/sympy/sympy/physics/control/lti.py", line 566, in doit
    a[i][j] = res.args[0][i][k] * arg.args[0][k][j]
TypeError: 'TransferFunction' object is not subscriptable
```
That's because `tfm2` is a column matrix, and also we need to make `[None, None]` (`(2, 3)` x `(3, 1)` -> `(2, 1)`) as the default list instead of `[[None], [None]]` which will then be passed into a `TransferFunctionMatrix` down below.

Similarly, we consider other cases...

Reply:
Why not always store the entries as a nested list, ie use `[[None, None]]`?

Reply:
> Why not always store the entries as a nested list, ie use `[[None, None]]`?

I don't understand how it's a column vector. Its shape would be (1, 2), no??